### DISCUSSION TITLE
Refreshing shellscope / changing site settings causes db mutations to have disposal exceptions?

### DISCUSSION THREAD

**Discussion by willnationsdev** [OrchardCMS/OrchardCore#9158] [Q&A] [answered]
**TLDR: How do I make sure that scoped services can safely use other dependency-injected scoped services (e.g. `IContentDefinitionManager`) without getting ObjectDisposedExceptions when the SiteSettings are modified? Or have I diagnosed this wrong?**

I want to have a shape template for a FlowPart that uses a specific stereotype. For some reason, this didn't appear to be built-in (just the content type). So in order to make it, I wrote an IShapeTableProvider that looked like this:

```cs
    public class MyStereotypeShapeTableProvider : IShapeTableProvider
    {
        private readonly IContentDefinitionManager _contentDefinitionManager;

        public MyStereotypeShapeTableProvider(IContentDefinitionManager contentDefinitionManager)
        {
            _contentDefinitionManager = contentDefinitionManager;
        }

        public void Discover(ShapeTableBuilder builder)
        {
            builder.Describe(nameof(FlowPart))
                .OnDisplaying(context =>
                {
                    dynamic dynamicShape = context.Shape;
                    FlowPart flowPart = dynamicShape.FlowPart;
                    var contentType = flowPart?.ContentItem?.ContentType;
                    if (string.IsNullOrEmpty(contentType))
                    {
                        return;
                    }
                    var typeDefinition = _contentDefinitionManager.GetTypeDefinition(contentType);
                    var settings = typeDefinition?.GetSettings<ContentTypeSettings>();
                    if (settings?.Stereotype == "MyStereotype" && flowPart.Widgets.Any())
                    {
                        context.Shape.Metadata.Wrappers.Add($"MyStereotype__{nameof(FlowPart)}");
                    }
                });
        }
    }
```

It works fine, but I have a script resource in the shape template. I experimented with toggling the UseCdn flag in the site settings GUI and suddenly loading the page now triggers an "object is disposed" error when fetching the definition. The issue goes away if I restart the application, but arises again if toggled once more.

Based on other issues/discussions, it seems that this is related to the `ISession` becoming disposed when the ShellScope is reloaded (which, as I understand it, happens when the site settings are changed). Is this the correct diagnosis? If so, why would my *scoped* service have a problem with this? Doesn't it get a fresh instance of the `IContentDefinitionManager` from *after* the shellscope refresh has already taken place? What are the general use cases I should identify and refactor to account for this issue? I would prefer to not have to restart my SaaS app every time a tenant changes their site settings.

**Comment by deanmarcussen** [comment, marked as answer]
You need to resolve the content definition manager from the describe context service provider, as it will have been disposed when constructor injected as you are currently doing

**Reply by willnationsdev** [reply]
Ah, I see. So the context has its own ServiceProvider. When do these kinds of things happen? When should I be on the look out for a ServiceProvider property in a context?

**Reply by deanmarcussen** [reply]
It tends to be Funcs which are described/built, and the the describe/builder is disposed of, while the Func stored by the building service.

 If that makes sense?

**Reply by willnationsdev** [reply]
I think I understand, yeah. So, if I'm working with any kind of Builder or "Describe" method of some kind, I should check the execution context to see if it has any `IServiceProvider` properties, and if so, defer to that rather than dependency injection. Sounds good. Thanks.
